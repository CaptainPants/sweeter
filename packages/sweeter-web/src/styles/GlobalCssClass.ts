import { preprocessClassContent } from './preprocessor/preprocess.js';
import type {
    GlobalStyleSheetContentGeneratorContext,
    AbstractGlobalCssStylesheet,
    StylesheetGenerator,
} from './types.js';

type Content =
    | ((self: GlobalCssClass) => string | StylesheetGenerator)
    | string
    | undefined;

export interface GlobalCssClassOptions {
    /**
     * Base the class name for this class on this value (it will be prefixed/suffixed or otherwise made unique).
     */
    className: string;
    content?: Content;
    preprocess?: boolean;
}

export class GlobalCssClass implements AbstractGlobalCssStylesheet {
    public readonly className: string;
    public readonly id: string;
    public readonly symbol: symbol;
    public readonly content?: Content;
    public readonly preprocess: boolean;

    constructor(options: GlobalCssClassOptions);
    constructor({
        className,
        content,
        preprocess = true,
    }: GlobalCssClassOptions) {
        this.className = className;
        this.id = className;
        this.symbol = Symbol('GlobalCssClass-' + className);
        this.content = content;
        this.preprocess = preprocess;
    }

    getContent(
        context: GlobalStyleSheetContentGeneratorContext,
    ): string | undefined {
        const name = context.getPrefixedClassName(this);

        let content =
            typeof this.content === 'string'
                ? this.content
                : this.content?.(this);

        if (!content) {
            return undefined;
        }

        if (typeof content === 'function') {
            content = content(context);
        }

        const transformed = this.preprocess
            ? preprocessClassContent(name, content)
            : `.${name}\r\n{\r\n${content}\r\n}`;
        return transformed;
    }
}
