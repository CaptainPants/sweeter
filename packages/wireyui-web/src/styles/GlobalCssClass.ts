import type { GlobalStyleSheet } from './types.js';

let counter = 0;

export class GlobalCssClass implements GlobalStyleSheet {
    public readonly className: string;
    public readonly content: string;
    public readonly id: symbol;

    constructor(options: { id: string; content: string }) {
        this.className = '_glbl' + counter + '_' + options.id;
        ++counter;

        this.content = `.${this.className} {\r\n${options.content}\r\n}`;
        this.id = Symbol('GlobalCssClass');
    }
}
