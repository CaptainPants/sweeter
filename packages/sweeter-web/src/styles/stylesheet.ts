import { type StylesheetGenerator } from './types.js';
import { type GlobalCssClass } from './GlobalCssClass.js';

/**
 * Template string function.
 *
 * Creates a StylesheetGenerator based on a templated string, where you can use GlobalCssClass references in style selectors.
 * @param template
 * @param params
 * @returns
 */
export function stylesheet(
    template: TemplateStringsArray,
    ...params: (GlobalCssClass | StylesheetGenerator | string)[]
): StylesheetGenerator {
    return (context) => {
        const res: string[] = [];
        const last = template.length - 1;
        for (let i = 0; i < template.length; ++i) {
            res.push(template[i]!);

            if (i !== last) {
                const param = params[i]!;

                if (typeof param === 'string') {
                    res.push(param);
                } else if (typeof param === 'function') {
                    res.push(param(context));
                } else {
                    res.push('.');
                    res.push(context.getPrefixedClassName(param));
                }
            }
        }
        return res.join('');
    };
}
