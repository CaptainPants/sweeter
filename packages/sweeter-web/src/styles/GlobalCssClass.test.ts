import { stylesheet } from './stylesheet.js';
import { GlobalCssClass } from './GlobalCssClass.js';

it('Common case produces sensible result', () => {
    const class_ = new GlobalCssClass({
        className: 'test',
        content: `
            color: white;
            background: black;
        `,
    });

    expect(
        class_.getContent({
            getPrefixedClassName: (class_) => class_.className,
        }),
    ).toMatchSnapshot();
});

it('Self referential', () => {
    const class_ = new GlobalCssClass({
        className: 'test',
        content: (self) => stylesheet`.NestedOnce { ${self} { color: green; } }`,
    });

    expect(
        class_.getContent({
            getPrefixedClassName: (class_) => class_.className,
        }),
    ).toMatchSnapshot();
});
