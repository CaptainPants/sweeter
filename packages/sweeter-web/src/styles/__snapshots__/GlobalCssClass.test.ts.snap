// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Common case produces sensible result 1`] = `
".test {
    color: white;
    background: black;
}
"
`;

exports[`Self referential 1`] = `
".test .NestedOnce .test {
    color: green;
}
"
`;
